internal/db/schema/migrations/oss/postgres/88/01_remove_session_connection_state.up.sql

```diff
@@ -0,0 +1,231 @@
+  -- "Closed" is handled by the trigger function, update_connected_time_range_on_closed_reason, which sets the connected_time_range upper bound to now()
+  -- State transitions are guarded by the trigger function, check_connection_state_transition, which ensures that the state transitions are valid
+
+  create or replace function check_connection_state_transition() returns trigger
+  as $$
+    begin
+      -- Authorized state
+      if new.connected_time_range is null then
+        return new;
+      end if;
+      -- If the old state was authorized, any transition is valid
+      if old.connected_time_range is null then
+        return new;
+      end if;
+      -- Prevent transitions from connected to connected
+      if upper(old.connected_time_range) = 'infinity' and upper(new.connected_time_range) = 'infinity' then
+        raise exception 'Invalid state transition from connected to connected';
+      end if;
+      -- Prevent transitions from closed to connected
+      if lower(new.connected_time_range) >= upper(old.connected_time_range) then
+        raise exception 'Invalid state transition from closed to connected';
+      end if;
+      return new;
+    end;
+  $$ language plpgsql;
+
+  create trigger check_connection_state_transition before update of connected_time_range on session_connection
+    for each row execute procedure check_connection_state_transition();
+
+  create or replace function update_connected_time_range_on_closed_reason() returns trigger
+  as $$
+    begin
+      if new.closed_reason is not null then
+        perform from
+          session_connection cs
+        where
+          cs.public_id = new.public_id and
+          -- If the connection is already closed, there's no need to update the connected_time_range
+          upper(cs.connected_time_range) <= now();
+        if not found then
+          update session_connection
+          set
+            connected_time_range = tstzrange(lower(connected_time_range), now())
+          where
+            public_id = new.public_id;
+        end if;
+      end if;
+    return new;
+    end;
+  $$ language plpgsql;
+
+  create trigger update_connected_time_range_closed_reason after update of closed_reason on session_connection
+    for each row execute procedure update_connected_time_range_on_closed_reason();
+
+  create or replace function update_session_state_on_termination_reason() returns trigger
+    as $$
+  begin
+    if new.termination_reason is not null then
+      perform  from
+        session
+      where
+        public_id = new.public_id and
+        public_id not in (
+            select session_id
+              from session_connection
+            where
+              -- open connections will have a connected_time_range with an upper bound of infinity
+              upper(connected_time_range) > now()
+        );
+      if not found then
+        raise 'session %s has open connections', new.public_id;
+      end if;
```

Comment:
I think this can be done without the subquery and without using `not in`:

```sql
perform
   from session_connection
  where session_id                  = new.public_id
    and upper(connected_time_range) = 'infinity'::timestamptz;
if found then
  raise 'session %s has open connections', new.public_id;
end if;
```